### PR TITLE
Harden WooCommerce event builder with defensive guards and cart quantity checks

### PR DESCRIPTION
**Pull request by bojanaivovic**
## Description

Hardens WooCommerce event construction paths to safely handle invalid or missing data and cart edge cases.

Includes defensive type checks for WooCommerce objects (e.g. product/order instances) before building event payloads, skips invalid order/cart line items when product data is missing or not a valid `WC_Product`, and adds quantity validation (`<= 0`) before using values in calculations. Also prevents division-by-zero in cart value and item price calculations by guarding item quantity before division.

These changes ensure only valid data is sent in events and prevent PHP warnings or fatal errors in edge cases.

### Type of change

- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have confirmed that my changes do not introduce any new PHPCS warnings or errors.
- [x] I have checked plugin debug logs that my changes do not introduce any new PHP warnings or FATAL errors.
- [x] I followed general Pull Request best practices. Meta employees to follow this wiki: https://fburl.com/wiki/b2b0midg
- [ ] I have added tests (if necessary) and all the new and existing unit tests pass locally with my changes.
- [x] I have completed dogfooding and QA testing, or I have conducted thorough due diligence to ensure that it does not break existing functionality.
- [ ] I have updated or requested update to plugin documentations (if necessary).

## Changelog entry

Add defensive WooCommerce event guards and cart quantity validation to avoid invalid payloads and division-by-zero edge cases.

## Test Plan

1. Purchase flow sanity check  
   - Place an order with standard WooCommerce products (including variation products)  
   - Confirm Purchase event payload is generated as expected  

2. Invalid product/order defensive checks  
   - Trigger event paths with missing/invalid product/order references  
   - Confirm no PHP warnings/fatals and invalid items are skipped safely  

3. Cart quantity edge cases  
   - Test cart lines with zero/invalid quantity  
   - Confirm no division-by-zero warnings and value/content calculations skip invalid lines  

4. Regression check  
   - Verify normal AddToCart/ViewContent/Purchase behavior remains unchanged for valid cart/order data

### CODE REVIEW

File: integration/class-facebookwordpresswoocommerce.php [hunk 1]
```diff
@@ -554,12 +568,19 @@ private static function getPiiFromBillingInformation( $order ) {
      *                    or null if the cart item is empty.
      */
     private static function getAddToCartValue( $cart_item, $quantity ) {
-        if ( ! empty( $cart_item ) ) {
-            $price = $cart_item['line_total'] / $cart_item['quantity'];
-            return $quantity * $price;
+        if ( empty( $cart_item )
+            || ! isset( $cart_item['quantity'] )
+            || ! isset( $cart_item['line_total'] ) ) {
+            return null;
```
Comment (vahidkay-meta) [inline integration/class-facebookwordpresswoocommerce.php:574]: Should this be null or 0 or throw an exception?

File: integration/class-facebookwordpresswoocommerce.php [hunk 2]
```diff
@@ -554,12 +568,19 @@ private static function getPiiFromBillingInformation( $order ) {
...
         }
 
-        return null;
+        $item_quantity = (int) $cart_item['quantity'];
+        if ( $item_quantity <= 0 ) {
+            return null;
```
Comment (vahidkay-meta) [inline integration/class-facebookwordpresswoocommerce.php:579]: same as above

File: integration/class-facebookwordpresswoocommerce.php [hunk 3]
```diff
@@ -622,19 +644,60 @@ private static function getContentIds( $cart ) {
...
+
+    /**
+     * Validates that the given value behaves like a WooCommerce order.
+     *
+     * @param mixed $order Order-like value to validate.
+     *
+     * @return bool Whether the value can be safely used as an order.
+     */
+    private static function isValidOrderObject( $order ) {
+        return is_object( $order )
```
Comment (vahidkay-meta) [inline integration/class-facebookwordpresswoocommerce.php:688]: Shouldn't we check if it's an object of type WC_Order, instead of this?

File: integration/class-facebookwordpresswoocommerce.php [hunk 4]
```diff
@@ -622,19 +644,60 @@ private static function getContentIds( $cart ) {
...
 
+    /**
+     * Validates that the given value behaves like a WooCommerce product.
+     *
+     * @param mixed $product Product-like value to validate.
+     *
+     * @return bool Whether the value can be safely used as a product.
+     */
+    private static function isValidProductObject( $product ) {
+        return is_object( $product )
```
Comment (vahidkay-meta) [inline integration/class-facebookwordpresswoocommerce.php:674]: Shouldn't we check if it's of WC_Product type?

File: integration/class-facebookwordpresswoocommerce.php (hunk 1 above)
Comment (bojanaivovic) [reply integration/class-facebookwordpresswoocommerce.php:574]: Hi @vahidkay-meta, great question.                                                                                                                   

In this context, null most accurately communicates that the value could not be determined from the cart item data. Returning 0 could be misinterpreted as a valid calculated amount, while throwing an exception would be unnecessarily disruptive for non-critical tracking flow.                                                                                                              
So from both a technical and product perspective, null is the safest and clearest choice here.

File: integration/class-facebookwordpresswoocommerce.php (hunk 2 above)
Comment (bojanaivovic) [reply integration/class-facebookwordpresswoocommerce.php:579]: That should also remain null for the same reason.                                                                                 
                                                                                                                                              
If quantity <= 0, we can’t compute a meaningful unit price (line_total / quantity), so the value is effectively invalid/unavailable. Returning 0 would again imply a valid computed amount, and throwing an exception would be too heavy for tracking logic.

File: integration/class-facebookwordpresswoocommerce.php (hunk 3 above)
Comment (bojanaivovic) [reply integration/class-facebookwordpresswoocommerce.php:688]: I agree that instanceof WC_Order / instanceof WC_Product is a stricter and more explicit check.                                              
That said, in the WooCommerce integration we currently use behavior-based validation (is_object + method_exists) rather than checking a concrete class type.                                                                                                                                        
                                                                                                                                              
The practical reason is that this logic only needs specific methods to be present, which keeps it compatible with wrapped/decorated WooCommerce objects and with test mocks (which are not necessarily instances of WooCommerce classes). The same pattern is used in this part of the integration, so this keeps the validation style consistent.                                     
                                                                                                                                              
If we want to move to instanceof, we can absolutely do that - but it should be done systematically (including test updates) to avoid mixing validation approaches.

**Comment by ukilla** [comment]
#171 rewrote the WooCommerce integration and picked up part of it: invalid products are now skipped in purchase items, `get_product_id()` is null safe, and the checkout contents check quantity. 

Three cases from this PR are still open on main:
- `get_purchase_event_data()` and `get_add_to_cart_event_data()` divide by quantity with no zero check, which is a fatal on PHP 8.
- `get_purchase_event_data()` calls `get_items()` on the order without checking the lookup returned one.
- `line_total` is read without `isset()` in the add-to-cart and checkout builders.

Maybe rebase onto main and narrow the PR to those three, with tests in `FacebookWordpressWooCommerceTest`? Then it should be good to go.

**Comment by bojanaivovic** [comment]
> #171 rewrote the WooCommerce integration and picked up part of it: invalid products are now skipped in purchase items, `get_product_id()` is null safe, and the checkout contents check quantity.
> 
> Three cases from this PR are still open on main:
> 
> * `get_purchase_event_data()` and `get_add_to_cart_event_data()` divide by quantity with no zero check, which is a fatal on PHP 8.
> * `get_purchase_event_data()` calls `get_items()` on the order without checking the lookup returned one.
> * `line_total` is read without `isset()` in the add-to-cart and checkout builders.
> 
> Maybe rebase onto main and narrow the PR to those three, with tests in `FacebookWordpressWooCommerceTest`? Then it should be good to go.

Thanks! Addressed in [commit](https://github.com/facebookincubator/Facebook-Pixel-for-Wordpress/pull/139/changes/56776ddf95223f2eb3c1010147a03384ad8f1a35).
